geo/ogrpolygon.cpp

```diff
@@ -0,0 +1,236 @@
+#include "ogrpolygon.hpp"
+#include "csconvertor.hpp"
+
+namespace geo::ogr
+{
+
+namespace fs = boost::filesystem;
+
+MultiPolygon::MultiPolygon(const OGRMultiPolygon& polygons)
+    : polygons_(polygons)
+{ }
+
+MultiPolygon::MultiPolygon(MultiPolygon&& other)
+    : polygons_(std::move(other.polygons_))
+{ }
+
+const OGRMultiPolygon& MultiPolygon::polygons() const
+{
+    return polygons_;
+}
+
+bool MultiPolygon::overlaps(const OGRMultiPolygon& other) const
+{
+    return polygons_.Overlaps(&other);
+}
+
+bool MultiPolygon::overlaps(const MultiPolygon& other) const
+{
+    return overlaps(other.polygons_);
+}
+
+bool MultiPolygon::overlaps(const math::Extents2& other) const
+{
+    OGRLinearRing ring;
+
+    math::Point2 p { ll(other) };
+    ring.addPoint(p[0], p[1]);
+    p = lr(other);
+    ring.addPoint(p[0], p[1]);
+    p = ur(other);
+    ring.addPoint(p[0], p[1]);
+    p = ul(other);
+    ring.addPoint(p[0], p[1]);
+    p = ll(other);
+    ring.addPoint(p[0], p[1]);
+
+    OGRPolygon polygon;
+    polygon.addRing(&ring);
+    return polygons_.Overlaps(&polygon);
+}
+
+bool MultiPolygon::overlaps(const math::Points2& other) const
+{
+    if (other.size() < 3) return false;
+    OGRLinearRing ring;
+    ring.addPoint(other[0][0], other[0][1]);
+    ring.addPoint(other[1][0], other[1][1]);
+    ring.addPoint(other[2][0], other[2][1]);
+    ring.addPoint(other[0][0], other[0][1]);
+
+    OGRPolygon polygon;
+    polygon.addRing(&ring);
+    return polygons_.Overlaps(&polygon);
+}
+
+MultiPolygon MultiPolygon::from(const math::Extents2& extents)
+{
+    return toPolygons(extents);
+}
+
+void convert(const CsConvertor& conv, OGRPolygon& polygon)
+{
+    for (auto& ring : polygon)
+    {
+        for (OGRPoint& p : ring)
+        {
+            math::Point2 p_conv = conv(math::Point2 { p.getX(), p.getY() } );
+            p.setX(p_conv[0]);
+            p.setY(p_conv[1]);
+        }
+    }
+}
+
+void convert(const CsConvertor& conv, OGRMultiPolygon& polygons)
+{
+    for (auto& polygon : polygons)
+    {
+        convert(conv, *polygon);
+    }
+}
+
+MultiPolygon toPolygons(const math::Extents2& extents)
+{
+    OGRMultiPolygon polygons;
+    OGRPolygon polygon;
+    OGRLinearRing ring;
+
+    math::Point2 p { ll(extents) };
+    ring.addPoint(p[0], p[1]);
+    p = lr(extents);
+    ring.addPoint(p[0], p[1]);
+    p = ur(extents);
+    ring.addPoint(p[0], p[1]);
+    p = ul(extents);
+    ring.addPoint(p[0], p[1]);
+    p = ll(extents);
+    ring.addPoint(p[0], p[1]);
+
+    polygon.addRing(&ring);
+    polygons.addGeometry(&polygon);
+    return { polygons };
+}
+
+GDALDatasetUniquePtr openDataset(const fs::path& path)
+{
+    GDALDatasetUniquePtr dataset {
+        GDALDataset::Open(reinterpret_cast<const char*>(path.c_str()), GDAL_OF_VECTOR | GDAL_OF_READONLY)
```

Comment:
@pato-melowntech why do we need the cast for Win?

Reply:
This should work, reinterpret cast does not make much sense.
```suggestion
        GDALDataset::Open(path.generic_string().c_str(), GDAL_OF_VECTOR | GDAL_OF_READONLY)
```